lib/ClangImporter/ImportMacro.cpp

```diff
@@ -424,120 +459,186 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
-                          clangTy->isUnsignedIntegerType() };
-      return createMacroConstant(impl, macro, name, DC, type,
-                                 clang::APValue(value),
-                                 ConstantConvertKind::Coerce, /*isStatic=*/false,
-                                 ClangN);
-    // Check for an expression of the form (FLAG1 | FLAG2), (FLAG1 & FLAG2),
-    // (FLAG1 || FLAG2), or (FLAG1 || FLAG2)
-    } else if (tokenI[0].is(clang::tok::identifier) &&
-               isBinaryOperator(tokenI[1]) &&
-               tokenI[2].is(clang::tok::identifier)) {
-      auto firstID = tokenI[0].getIdentifierInfo();
-      auto secondID = tokenI[2].getIdentifierInfo();
-
-      if (firstID->hasMacroDefinition() && secondID->hasMacroDefinition()) {
-        auto firstMacroInfo = impl.getClangPreprocessor().getMacroInfo(firstID);
-        auto secondMacroInfo = impl.getClangPreprocessor().getMacroInfo(
-                                                                      secondID);
-        auto firstIdentifier =
-            impl.getNameImporter().importMacroName(firstID, firstMacroInfo);
-        auto secondIdentifier =
-            impl.getNameImporter().importMacroName(secondID, secondMacroInfo);
-        impl.importMacro(firstIdentifier, firstMacroInfo);
-        impl.importMacro(secondIdentifier, secondMacroInfo);
-        auto firstIterator = impl.ImportedMacroConstants.find(firstMacroInfo);
-        if (firstIterator == impl.ImportedMacroConstants.end()) {
-          return nullptr;
-        }
-        auto secondIterator = impl.ImportedMacroConstants.find(secondMacroInfo);
-        if (secondIterator == impl.ImportedMacroConstants.end()) {
-          return nullptr;
-        }
+    // The final value of the constant.
+    llvm::APSInt resultValue;
+    // The result's APSInt specification and corresponding Swift type,
+    // after sign & width resolution.
+    std::pair<clang::ento::APSIntType, Type>
+      resultType = { resultValue, nullptr };
+
+    auto firstBitWidth  = firstValue.getBitWidth();
+    auto secondBitWidth = secondValue.getBitWidth();
+
+    // Ensure both values have the same signedness.
+    if (firstValue.isSigned() != secondValue.isSigned()) {
+
+      // If unsignedTy >= signedTy, convert signed -> unsigned.
+      if (firstValue.isUnsigned() && firstBitWidth >= secondBitWidth) {
+        resultType = { firstValue, firstType };
+      } else if (secondValue.isUnsigned() && secondBitWidth >= firstBitWidth) {
+        resultType = { secondValue, secondType };
+
+      // If signedTy can represent all unsigneds, convert unsigned -> signed.
+      } else if (firstValue.isSigned() && firstBitWidth > secondBitWidth) {
+        resultType = { firstValue, firstType };
+      } else if (secondValue.isSigned() && secondBitWidth > firstBitWidth) {
+        resultType = { secondValue, secondType };
+      } else {
+        llvm_unreachable("Types have unequal bitwidths, yet neither is greater "
```

Comment:
What about when the first value is signed but the first bit width is smaller than the second?